### DISCUSSION TITLE
Zero Termination Fees for CC Sectors

### DISCUSSION THREAD

**Discussion by The-Wayvy** [filecoin-project/FIPs#972] [Enhancements - Technical]
**Summary**

Set the termination fee of CC sectors to zero.

**Motivation**

Encourage new miners to try Filecoin (lower barrier to entry)

Decrease risk for all miners (pledge becomes a fully liquid asset)

**Rationale**

Other chains allow miners to join and leave at will. 

Duration requirements handicap Filecoin in the competition to attract hardware/capital without providing any benefits to compensate.  

**References**

[Discussion 691](https://github.com/filecoin-project/FIPs/discussions/691)

[Discussion 712](https://github.com/filecoin-project/FIPs/discussions/712)

**Comment by Fatman13** [comment]
LGTM

**Comment by steven004** [comment]
A fair proposal.

I support this, aside from zero termination fee for CC sectors, I also advocate for waiving the fault penalty for CC sectors.

To further this, I am particularly keen to simplify the layer 1 protocol and overhaul the QAP design. This involves dividing the current consensus into a combination of PoPW (rawbytes storage) and PoS (collateral), as separate entities, akin to the model outlined in https://github.com/filecoin-project/FIPs/discussions/442.

**Comment by Stebalien** [comment]
Miners in Bitcoin can come and go because all they're doing is sequencing messages. On the other hand, in Filecoin, storage providers are providing long-term storage for the network and are generally expected to stick around and be online to provide that storage.

My concern is that this proposal (as-is) makes storing user data less attractive when compared to CC sectors as storing user data is now strictly riskier and less flexible. CC sectors are only rewarded _at all_ because they demonstrate committed capacity to the network. They're demonstrating that the SP _could_ use this capacity to store user data.

A middle ground might be a 0 or significantly reduced termination fee for non-faulty CC sectors, with some kind of ramp down delay. That is:

1. The SP submits a `ProposeTermination` message, marking a set of CC sectors as pending termination.
2. If, after proposing termination, the sector becomes faulty, this mark is removed and the sector is no longer proposed for termination.
3. Finally, some number of epochs later, the SP can submit a `FinishTermination` message, completely terminating these sectors (with a reduced and/or zero fee).

That way these sectors can still proved that the SP can reliably store data.

The one missing piece here is a demonstration of long-term commitment, which is important for storing user data. One solution there is to have a minimum sector duration where the sector must live for, for example, six months before it can have this zero or low termination fee.

Additionally, I'd like to explore other ramp down solutions such as selling sectors (even sectors with user data) to other SPs.

**Reply by The-Wayvy** [reply]
**Allowing CC miners to terminate without penalty doesn't make them any less capable of storing user data.** 

Nor does requiring that they broadcast a message and wait a certain number of epochs before terminating make them more capable of storing user data. 

IMO this is a very weak argument in favor of duration requirements

~

The extra risk and loss of flexibility that data deals entail will be made worth it by real paying users.

If such a user is not to be found, then we still want miners to add hardware to Filecoin, and that means we shouldn't make miners' lives harder than necessary since we are competing for people/capital with other protocols.  

~

IMO The real risk CC miners face from termination fees far outweighs the imagined benefits to users. 

~

I agree with you about exploring hand-offs for deal sectors to avoid termination fees.

~

More generally I'd like to see a less prescriptive protocol. Leave deal duration, sector duration, etc. up to users and builders.

**Reply by The-Wayvy** [reply]
What matters for users is the number of miners online at any given moment, not how long any individual miner is committed to staying online . What’s better for users, one miner committed to 1000 hours or 1000 miners committed to 1 hour?

**Reply by rjan90** [reply]
> What matters for users is the number of miners online at any given moment, not how long any individual miner is committed to staying online . What’s better for users, one miner committed to 1000 hours or 1000 miners committed to 1 hour?

Its not a matter of choosing one or the other. The network wants many SPs, and long term oriented SPs.

**Reply by The-Wayvy** [reply]
As long as there are strong incentives or guarantees that deal data isn't lost, the number of "long term" people shouldn't matter.

~

a termination fee for CC miners incentivizes duration per miner at the expense of  number of miners.

I don't think this is justified at the consensus layer.

**Reply by rjan90** [reply]
> a termination fee for CC miners incentivizes duration per miner at the expense of number of miners.

I have not advocated for having a termination fee for CC-sectors in my previously reply, so please do not use that as an argument against my reply. I merly stated that we should endorse having many SPs in the network that are long-term oriented!

I think the Zero Termination Fee for CC Sector proposal is a good way to reduce the entry barriers for storage providers. But Steb raises important points we should think about, and try to understand to ensure we don't encourage short-term strategies that could also impact network security negatively.

**Reply by Fatman13** [reply]
> Miners in Bitcoin can come and go because all they're doing is sequencing messages. On the other hand, in Filecoin, storage providers are providing long-term storage for the network and are generally expected to stick around and be online to provide that storage.

I think many see that Filecoin is a **platform** as laid out in [Filecoin Mission](https://github.com/filecoin-project/FIPs/blob/master/mission.md#foundation), in which the core protocol doesn't force itself upon the relationship between a storage client and provider(s) but rather provide the support framework for user defined VM based storage market(s) to thrive. In such free market(s), storage clients would like to have their own storage terms in regard to pledge, penalty & rewards with its provider(s) not terms dictated by the core protocol. The more core protocol is entangled with storage, the harder a true market will emerge. What's worse, core protocol right now is driving away consensus (RBP) too.

**Reply by taoshengshi** [reply]
At the key of this matter is a tragedy of the Commons, a struggle for profit between miners and the public network. Are users and developers involved?

**Reply by laudiacay** [reply]
heavy +++++ on selling sectors

**Reply by laudiacay** [reply]
(AA could work for this...? steb idk)

**Reply by Stebalien** [reply]
SPs that aren't interested or able to reliably store user data are extracting rewards that should go to SPs who are, that's why I care about this. I do want to enable clients and SPs to work together in any way they see fit, but I also want to ensure they're _rewarded_ for it as much as possible: the block reward _exists_ to subsidize this interaction.

In terms of my proposal...

1. I proposed a ramp down delay to ensure only "healthy" sectors can be terminated free of charge (faults cannot be free). Unfortunately this requires a small delay because WindowedPoSt is probabilistic.
2. I don't have strong opinions on my minimum duration proposal. Honestly, it probably isn't necessary given a ramp down delay, users can and likely will prefer SPs with longer track records anyways.

**Reply by Fatman13** [reply]
> the block reward exists to subsidize this interaction.

imo, the block reward exists to encourage new network participants to join & to allow the network to grow, but right now the skewed incentive is doing the exact opposite (decrease of RBP) for well over a year.

> I proposed a ramp down delay to ensure only "healthy" sectors can be terminated free of charge (faults cannot be free).

imo, this is the exact kind of core protocol complexity (over-engineering) which are preventing new investment / capital into the network, sprinkles on top of already bloated core protocol.

**Comment by anorth** [comment]
Thanks for beginning this discussion. I'm broadly supportive and would like to help figure out the details. I think we want to reach a scheme that's as simple as possible without breaking anything significant.

### What's a CC sector?

For clarity, I am taking "CC sector" to mean a sector that is provably empty of any data. We can prove this contents by inspecting the data commitment (unsealed CID). After FIP-0076, CC is not the same as "sector without deals" or "sector with FIL+ that expired". The only thing we can simply and reliably test on-chain is "commits to zero data". Once we implement re-snap, it will be possible to transition a sector from "has some data" to "has zero data", but note that that's not possible today.

### What fee is reasonable?

I agree with the motivations to reduce/remove termination fees. However I don't think exactly zero is the right number. As I surveyed in https://github.com/filecoin-project/FIPs/discussions/691#discussioncomment-5749562, I think there are _some_ reasonable motivations for a non-zero offboarding fee. The most reasonable is to incentivize some stability of pledge (or storage, whichever you want to think secures the network). Complete freedom to add/remove network power at any rate at any time ("flash") makes it unreasonably difficult to ensure incentive alignment.

Almost all other chains, and on-chain applications, except Bitcoin limit the rate of on/off-boarding of pledge ("churn"). They usually do this with a lockup delay, rather than a fee. Lockups are an awkward fit for Filecoin sectors, though, because if an SP has lost the ability or desire to prove a sector, we can't enforce that they keep doing it anyway, whereas a proof-of-stake chain can enforce a token lockup period. 
1. We could consider designing a lockup period for the pledge, such that CC termination is free but the return of pledge is delayed. I don't necessarily oppose this, but it adds more complexity.
2. A lockup period for the sectors themselves (like @Stebalien's proposal) is even more complex, and kinda reduces to a termination fee for sectors that aren't proven during the lockup. The protocol must handle physical loss of sectors, but cannot distinguish that from a voluntary withdrawal.
3. IMO it would be simpler to translate the lockup period into a financially-similar fee, by burning `notional-lockup-period * notional-interest-rate * pledge` tokens immediately on off-boarding. This would be much less than the current termination fees. A ballpark might be `30 days * 15%p.a = 1.23%`; the current CC termination fee is a little over 8% of pledge. Maybe the notional period could be shorter.

I would note that actually delaying the release of pledge might have concrete security benefits over a 1% fee, the cost to an attacker with locked pledge being >>1%. I could easily be convinced the lockup is better, and worth the complexity. And maybe a delay is preferable to a fee to the good-faith SP operators anyway?

### Further simplifications

If we touch anything related to termination fees, I think we should take the opportunity to make other easy-to-reach simplifications too. We can make the fees for all sectors (not just CC) much simpler, much more efficient, and remove onboarding-epoch bias.

Following the [design goals sketched here](https://github.com/filecoin-project/FIPs/discussions/691#discussioncomment-5775777), I think we should:
- Stop discriminating between sectors based on activation epoch, and specify the termination fee per unit power in terms of current rewards. E.g. `CapacityOffboardingFee = 1.23% * ConsensusPledge` (or zero, if we do lock-up instead) and `DataTerminationFee = 90 * share-of-current-daily-block-reward`. A data sector terminated would pay both fees. 

This would permit us to remove 20% of active chain state and some code, and make offboarding costs more uniform for all SPs.

(Note that ConsensusPledge does depend on onboarding epoch. I think we should resolve that separately, e.g. #734 or something larger)

**Reply by The-Wayvy** [reply]
ANorth, thanks for the attention you’ve given this topic.

Your [summary](https://github.com/filecoin-project/FIPs/discussions/691#discussioncomment-5749562) of the arguments in favor of termination fees is excellent.

~

Based on your post, I have two questions:

1.) Is there a security threat from zero termination fees and lock-up? 

2.) Does anyone benefit from pledge and storage stability?

If the answer to both of these questions is “No”, then IMO we should go with zero termination fees / lock-up.

**Reply by The-Wayvy** [reply]
As for stability somehow being good for our ‘image’, you could make an argument in the other direction: “18 months of losing capacity, flatlined QAP, and decreasing number of independent entities securing Filecoin is _horrible_ for our image. We should do whatever we can to reverse these trends, even at the cost of higher storage/pledge volatility.”

**Reply by The-Wayvy** [reply]
Imagine someone proposed increasing termination fees to 100% of pledge (for the sake of stability). 
Would anyone support this? I don't think so.

The strength of crypto is precisely that the protocol is agnostic to which miner is running it. 

“long term commitment” is superfluous, and pursuing it at the expense of ease of participation only hurts us in the competition for miners/stakers.

I don’t support maintaining costs and risks to participation beyond what is absolutely necessary for the security of the network.

**Reply by anorth** [reply]
**Security:** Yes, I believe there is some security risk from zero termination fees or lock period. I'm not really the right person to quantify it, but I think we need some amount of either one or the other. 
- A fee is a simpler change from where we are now, and maintains security similar to present
- A lock-up of the pledge is probably _better_ for security, and could then permit zero fee. It might be simple, but needs a little more thought how to implement it (can we re-use the existing vesting table?).

The more I think about it, the more appealing a delayed release of the full pledge becomes. It's probably worth investigating, I will ping some people.

**Stability:** Beyond it's security impact, I tend to agree that further stability (of CC sectors) is probably not something the protocol should sacrifice much for. Removing costs and barriers is probably preferable. But this is a point of view others can legitimately disagree with. As noted #691, I also place very little value on the notion of commitment to future behaviour.

**Reply by lucaniz** [reply]
@The-Wayvy, regarding this
> 1.) Is there a security threat from zero termination fees and lock-up?

termination fees is needed to ensure rationality of storage (specifically for WindowPoSt security). The constraint we have is Termination Fee needs to be greater or equal than Fault Fee in order to prevent a strategy for which terminating a faulty sector is more rational than recovering it.

You can find details in the full analysis (section 5.1) that @irenegia and I put together here:

https://drive.google.com/file/d/1notObdkPT1BCztgspIpzSUAzWSrM8h81/view

**Reply by Fatman13** [reply]
> https://drive.google.com/file/d/1notObdkPT1BCztgspIpzSUAzWSrM8h81/view

The premise that the analysis (the paper above) is based on is the assumption of the existence of a storage market. But look around, there is no storage market on the network. The paper above is not even stored on Filecoin, its on Google drive.

**Reply by The-Wayvy** [reply]
@lucaniz 

If 

~ termination fee is equal to fault fee
~ and we require miners broadcast a message saying they intend to terminate, then submit one more window PoST, and only then terminate [irene](https://filecoinproject.slack.com/archives/C04USV0L7PF/p1712763577254839?thread_ts=1712262360.495279&channel=C04USV0L7PF&message_ts=1712763577.254839)


do we still need to delay return of pledge to the terminating miner? (this was proposed as an alternative to termination fees)

in other words, would finality or liveness guarantees or PoST security be affected if we followed the first two ~points in this post and returned all miner collateral instantly upon successful termination?

**Reply by anorth** [reply]
I believe PoST is secure if we do only those first two points, but that this is not the whole story of consensus security. I think we still should delay the release of collateral further for reasons analogous to why PoS networks implement lockups or exit queues. This will be both cheaper for SPs (even accounting for interest on funds) and an improvement in security from the current termination fees.

If 24hr is the ideal lockup, maybe these reduce to the same thing. But that's probably too short.

**Reply by The-Wayvy** [reply]
Then how should we decide the length of the lock-up period?

@anorth @irenegia

**Reply by irenegia** [reply]
24h if okay for windowPoSt, but even higher values are fine (ie, 24h is the minimum). So we can set it to whatever is needed for lockup, maybe setting to 1 week is okay for that?

**Reply by The-Wayvy** [reply]
So are we going to choose an arbitrary amount of time greater than or equal to 24 hours?

**Reply by The-Wayvy** [reply]
I'm assuming there is no formal analysis available of lock-up duration and consensus security that we could use to make a decision

**Reply by anorth** [reply]
Yes, I think you're right. We're handwaving a somewhat arbitrary "you can't attack the network without risking tokens for X days after terminating/exiting". I would start with 10-14 days, but yes it's somewhat arbitrary.

**Reply by The-Wayvy** [reply]
@anorth 

Is this FIP ready for Last Call?

I have no objection to starting with a 14 day lock-up.

**Reply by rjan90** [reply]
> Is this FIP ready for Last Call?

I don´t think there is a FIP Draft open for this discussion yet. So first step would be to open a WIP/Draft PR for this change before it can be moved to Last Call, following the template here: https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0001.md#fip-header-preamble. Another example you can look at is this PR: https://github.com/filecoin-project/FIPs/pull/993.

**Reply by anorth** [reply]
Theres an as-yet unspecified details of exactly how SPs will send "a message saying they intend to terminate". Equivalently, how to implement a delayed termination. I'm afraid I don't have the bandwidth to figure this out myself right now, but we'll happily review proposals. It will probably involve something in the miner deadline accounting state.
